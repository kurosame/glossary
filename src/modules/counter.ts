import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const ADD_COUNT = 'ADD_COUNT'
export const ADD_SAGA_COUNT = 'ADD_SAGA_COUNT'
export const GET_SAGA_COUNT = 'GET_SAGA_COUNT'

export interface ICounterActions {
  addCount: ActionFunctionAny<Action<{}>>
  addSagaCount: ActionFunctionAny<Action<{ sagaCount: number }>>
  getSagaCount: ActionFunctionAny<Action<{}>>
}

export interface ICounterState {
  count: number
  sagaCount: number
}

const initialState: ICounterState = {
  count: 0,
  sagaCount: 0
}

export const { addCount, addSagaCount, getSagaCount } = createActions(
  ADD_COUNT,
  ADD_SAGA_COUNT,
  GET_SAGA_COUNT
)

export const counter: Reducer = handleActions(
  {
    [ADD_COUNT]: (state: ICounterState): ICounterState => {
      return { ...state, count: state.count + 1 }
    },
    [ADD_SAGA_COUNT]: (
      state: ICounterState,
      action: Action<ICounterState>
    ): ICounterState => {
      return {
        ...state,
        sagaCount:
          state.sagaCount + (action.payload ? action.payload.sagaCount : 0)
      }
    }
  },
  initialState
)
