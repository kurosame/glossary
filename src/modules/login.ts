import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const SET_IS_LOGIN = 'SET_IS_LOGIN'

export interface LoginActions {
  setIsLogin: ActionFunctionAny<Action<{ isLogin: boolean }>>
}

export interface LoginState {
  isLogin: boolean
}

const initialState: LoginState = { isLogin: false }

export const { setIsLogin } = createActions<{ isLogin: boolean }>(SET_IS_LOGIN)

export const login: Reducer<
  LoginState,
  Action<{ isLogin: boolean }>
> = handleActions(
  {
    [SET_IS_LOGIN]: (
      _state: LoginState,
      action: Action<{ isLogin: boolean }>
    ): LoginState => {
      return { isLogin: action.payload.isLogin }
    }
  },
  initialState
)
