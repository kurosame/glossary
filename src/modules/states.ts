import type { LoginState } from '@/modules/login'
import type { WordState } from '@/modules/word'

export interface States {
  login: LoginState
  words: WordState[]
}
