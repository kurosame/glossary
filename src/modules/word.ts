import { Reducer } from 'redux'
import {
  Action,
  ActionFunctionAny,
  createActions,
  handleActions
} from 'redux-actions'

// Action types
export const GET_WORDS = 'GET_WORDS'
export const SET_WORDS = 'SET_WORDS'

export interface IWordActions {
  getWords: ActionFunctionAny<Action<{}>>
  setWords: ActionFunctionAny<Action<{ words: IWordState[] }>>
}

export interface IWordState {
  id: string
  category: string
  titles: string[]
  description: string
}

const initialState: IWordState[] = []

export const { getWords, setWords } = createActions(GET_WORDS, SET_WORDS)

export const words: Reducer<IWordState[], Action<IWordState[]>> = handleActions(
  {
    [SET_WORDS]: (
      state: IWordState[],
      action: Action<IWordState[]>
    ): IWordState[] => {
      return !!action.payload.length ? action.payload : state
    }
  },
  initialState
)
