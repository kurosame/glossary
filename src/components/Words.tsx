import Word from '@/components/Word'
import { IWordState } from '@/modules/word'
import GridList from 'material-ui/GridList'
import React from 'react'

interface IProps {
  words: IWordState[]
}

export default class Words extends React.Component<IProps> {
  public render() {
    return (
      <GridList cols={1}>
        {this.props.words.map(w => <Word key={w.id} word={w} />)}
      </GridList>
    )
  }
}
