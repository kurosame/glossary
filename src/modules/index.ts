import { ICounterActions, ICounterState } from '@/modules/counter'
import { IWordActions, IWordState } from '@/modules/word'

export interface IActions {
  counter: ICounterActions
  word: IWordActions
}

export interface IStates {
  counter: ICounterState
  words: IWordState[]
}
