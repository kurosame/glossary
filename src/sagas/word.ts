import { collection, getDocs } from 'firebase/firestore'
import type { Action } from 'redux-actions'
import type { CallEffect, PutEffect, TakeEffect } from 'redux-saga/effects'
import { call, put, take } from 'redux-saga/effects'

import { firestore } from '@/firebase/index'
import type { WordState } from '@/modules/word'
import { GET_WORDS, setWords } from '@/modules/word'

export const getFirestoreWords = (): Promise<WordState[]> =>
  getDocs(
    collection(firestore(), 'words').withConverter<WordState>({
      fromFirestore: (s, o) => ({
        id: s.data(o).id,
        category: s.data(o).category,
        titles: s.data(o).titles,
        description: s.data(o).description
      }),
      toFirestore: w => ({ id: w.id, category: w.category, titles: w.titles, description: w.description })
    })
  )
    .then(res =>
      res.docs.map(d => ({
        id: d.id,
        category: d.data().category,
        titles: d.data().titles,
        description: d.data().description
      }))
    )
    .catch((err: Error) => {
      console.error(`GET_WORDS Firestore response error: ${err.message}`)
      return []
    })

export function* getWords(): IterableIterator<TakeEffect | CallEffect | PutEffect<Action<{ words: WordState[] }>>> {
  while (true) {
    yield take(GET_WORDS)
    yield put(setWords!({ words: yield call(getFirestoreWords) }))
  }
}
