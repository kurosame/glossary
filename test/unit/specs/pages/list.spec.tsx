import Words from '@/components/Words'
import ConnectToList, { List } from '@/pages/List'
import { States } from '@/modules/states'
import { WordState } from '@/modules/word'
import { mount, shallow } from 'enzyme'
import React from 'react'
import configureStore from 'redux-mock-store'

const state: { words: WordState[] } = {
  words: [
    {
      id: 'Vue.js',
      category: 'vue',
      titles: ['Vue', 'Vue.js'],
      description: 'It a Vue.js'
    },
    {
      id: 'React',
      category: 'react',
      titles: ['React'],
      description: 'It a React'
    }
  ]
}
const states: States = { login: { isLogin: false }, words: state.words }
const actions: any = {}

beforeEach(() => {
  actions.getWords = jest.fn()
})
afterEach(() => {
  actions.getWords.mockReset()
})

test('Call the getWords when words is nothing', () => {
  shallow(
    <List
      state={{ login: { isLogin: false }, words: [] }}
      actions={actions}
      match={{ params: { category: '' } }}
    />
  )

  expect(actions.getWords).toBeCalled()
})

test('Not call the getWords when words exists', () => {
  shallow(
    <List
      state={{ login: { isLogin: false }, ...state }}
      actions={actions}
      match={{ params: { category: '' } }}
    />
  )

  expect(actions.getWords).not.toBeCalled()
})

test('Elements of the Words when category is nothing', () => {
  const wrapper = mount(
    <ConnectToList match={{ params: { category: '' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(
    wrapper
      .find(Words)
      .first()
      .text()
  ).toEqual('Vue.jsVueVue.jsIt a Vue.jsReactReactIt a React')
})

test('Elements of the Words when category exists', () => {
  const wrapper = mount(
    <ConnectToList match={{ params: { category: 'react' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(
    wrapper
      .find(Words)
      .first()
      .text()
  ).toEqual('ReactReactIt a React')
})

test('Elements of the Words when category not exists', () => {
  const wrapper = mount(
    <ConnectToList match={{ params: { category: 'angular' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(
    wrapper
      .find(Words)
      .first()
      .text()
  ).toEqual('')
})

test('Match the snapshot when category is nothing', () => {
  const wrapper = shallow(
    <ConnectToList match={{ params: { category: '' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(wrapper.html()).toMatchSnapshot()
})

test('Match the snapshot when category exists', () => {
  const wrapper = shallow(
    <ConnectToList match={{ params: { category: 'react' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(wrapper.html()).toMatchSnapshot()
})

test('Match the snapshot when category not exists', () => {
  const wrapper = shallow(
    <ConnectToList match={{ params: { category: 'angular' } }} />,
    {
      context: {
        store: configureStore()(states)
      }
    }
  )

  expect(wrapper.html()).toMatchSnapshot()
})
