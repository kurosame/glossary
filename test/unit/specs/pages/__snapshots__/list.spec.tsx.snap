// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match the snapshot category exists 1`] = `"<ul class=\\"MuiList-root MuiList-padding\\" data-test=\\"words\\"><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"MuiPaper-root sc-AxjAm hmMezx MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline sc-AxirZ dzevfP\\" data-test=\\"search-bar-input\\"><input type=\\"text\\" placeholder=\\"Search...\\" class=\\"MuiInputBase-input MuiInput-input\\"/></div><span class=\\"material-icons MuiIcon-root sc-AxiKw hhXvMk\\" aria-hidden=\\"true\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span></div></li><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"sc-AxhUy dWeUmw\\"><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiCardHeader-root\\" data-test=\\"id\\"><div class=\\"MuiCardHeader-content\\"><span class=\\"MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock\\"><h3>React</h3></span></div></div><div class=\\"MuiCardContent-root\\" data-test=\\"titles\\"><span class=\\"sc-AxhCb bqDdOh\\"><div class=\\"MuiChip-root\\"><span class=\\"MuiChip-label\\">React</span></div></span></div><div class=\\"MuiCardContent-root\\" data-test=\\"description\\"><p>It a React</p></div></div></div></li></ul>"`;

exports[`Match the snapshot category is nothing 1`] = `"<ul class=\\"MuiList-root MuiList-padding\\" data-test=\\"words\\"><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"MuiPaper-root sc-AxjAm hmMezx MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline sc-AxirZ dzevfP\\" data-test=\\"search-bar-input\\"><input type=\\"text\\" placeholder=\\"Search...\\" class=\\"MuiInputBase-input MuiInput-input\\"/></div><span class=\\"material-icons MuiIcon-root sc-AxiKw hhXvMk\\" aria-hidden=\\"true\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span></div></li><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"sc-AxhUy dWeUmw\\"><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiCardHeader-root\\" data-test=\\"id\\"><div class=\\"MuiCardHeader-content\\"><span class=\\"MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock\\"><h3>Vue.js</h3></span></div></div><div class=\\"MuiCardContent-root\\" data-test=\\"titles\\"><span class=\\"sc-AxhCb bqDdOh\\"><div class=\\"MuiChip-root\\"><span class=\\"MuiChip-label\\">Vue</span></div></span><span class=\\"sc-AxhCb bqDdOh\\"><div class=\\"MuiChip-root\\"><span class=\\"MuiChip-label\\">Vue.js</span></div></span></div><div class=\\"MuiCardContent-root\\" data-test=\\"description\\"><p>It a Vue.js</p></div></div></div></li><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"sc-AxhUy dWeUmw\\"><div class=\\"MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiCardHeader-root\\" data-test=\\"id\\"><div class=\\"MuiCardHeader-content\\"><span class=\\"MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock\\"><h3>React</h3></span></div></div><div class=\\"MuiCardContent-root\\" data-test=\\"titles\\"><span class=\\"sc-AxhCb bqDdOh\\"><div class=\\"MuiChip-root\\"><span class=\\"MuiChip-label\\">React</span></div></span></div><div class=\\"MuiCardContent-root\\" data-test=\\"description\\"><p>It a React</p></div></div></div></li></ul>"`;

exports[`Match the snapshot category not exists 1`] = `"<ul class=\\"MuiList-root MuiList-padding\\" data-test=\\"words\\"><li class=\\"MuiListItem-root MuiListItem-gutters\\"><div class=\\"MuiPaper-root sc-AxjAm hmMezx MuiPaper-elevation1 MuiPaper-rounded\\"><div class=\\"MuiInputBase-root MuiInput-root MuiInput-underline sc-AxirZ dzevfP\\" data-test=\\"search-bar-input\\"><input type=\\"text\\" placeholder=\\"Search...\\" class=\\"MuiInputBase-input MuiInput-input\\"/></div><span class=\\"material-icons MuiIcon-root sc-AxiKw hhXvMk\\" aria-hidden=\\"true\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\"><path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"></path></svg></span></div></li></ul>"`;

exports[`Match the snapshot isLogin is false 1`] = `""`;
